Add release notes and blog post for 0.21.6 (#246)
TeamNewPipe/website: _posts/2020-07-13-newpipe-0.21.6-released.md

<!-- more -->

## New

You can now tap on hashtags in the stream description, which will execute a search with that hashtag. This was added by [@TiA4f8R](https://github.com/TiA4f8R).

[@mauriciocolli](https://github.com/mauriciocolli) introduced a new button which lets you hide videos you've already played within a feed! Simple in theory, but quite complex in execution. Eventually, [@Stypox](https://github.com/Stypox) took over the PR and finished ironing out the kinks, so enjoy your neat feeds devoid of previously seen videos!

[@Dakkaron](https://github.com/Dakkaron) added a new <code>Appearance</code> setting to manually enable or disable Tablet Mode. So if you have an Android TV which NewPipe mistakenly considers a phone or vice-versa, you can fiddle with this setting, so it works to your liking.

### Improved

> Full SAF ([Storage Access Framework](https://developer.android.com/guide/topics/providers/document-provider)) support is finally here! This makes NewPipe future-proof with respect to data storage and retrieval. Why? In Android 10, Google introduced a security feature known as Scoped Storage, which only lets an app see the specific folder or folders in the filesystem which pertain to it, and nothing else (such as your private data from another app). This requires SAF to be enabled. If that wasn't enough, SAF enables storing your downloads on external storage! This massive undertaking was initiated over a year ago by [@wb9688](https://github.com/wb9688), and was then taken on by [@Stypox](https://github.com/Stypox), who improved upon it. Thanks to both of them for getting this implemented! SAF will automatically be enabled for you if you have Android 6 and above, and is now the only option on Android 10 and above. Meanwhile, Kitkat users can only use the native file picker. Note: SAF is still disabled on Fire TV due to a Fire OS GUI bug.

Author: Why do you want to remove the now? We supported SAF before, but also allowed using the other file picker on Android 10 and above in prior NewPipe versions.

> NewPipe has learned to detect deleted channels, both upon opening them from the Subscriptions list, and when refreshing a feed containing such channels! Upon encountering such channels during refresh, the app will notify you and prompt you to delete them. This functionality required a small addition to NewPipe Extractor as well. [@TobiGr](https://github.com/TobiGr) took care of both sides, with help from [@Stypox](https://github.com/Stypox) regarding implementation. So no more snackbar errors every time you update a feed which has deleted channels! Moreover, opening a terminated channel will show you a proper error message explaining why the channel was terminated, instead of a generic error.

Reviewer: ```suggestion
NewPipe has learned to detect deleted channels, both upon opening them from the Subscriptions list, and when refreshing a feed containing such channels! Upon encountering such channels during  a refresh, the app will notify you and prompt you to delete them. This functionality required a small addition to NewPipe Extractor as well. [@TobiGr](https://github.com/TobiGr) took care of both sides, with help from [@Stypox](https://github.com/Stypox) regarding implementation. So no more snackbar errors every time you update a feed which has deleted channels! Moreover, opening a terminated channel will show you a proper error message explaining why the channel was terminated, instead of a generic error.
```

Everyone who enjoys reading software licenses (who doesn't?!) and does this inside the NewPipe app can now explore the websites of these licenses using the new "Open website" button in the license pop-up easily, thanks to [@Peyman-hme](https://github.com/Peyman-hme).

> [@TiA4f8R](https://github.com/TiA4f8R) upgraded the Share Sheet on Android 10 and above to show the title of the media/channel/playlist being shared. Some OEM ROMs or custom ROMs don't provide the "Copy to clipboard" option in the Share Sheet. So [@Abanoub8](https://github.com/Abanoub8) added the ability to long-press the Share button below the player, which copies the media URL directly into your clipboard.

Author: ```suggestion
[@TiA4f8R](https://github.com/TiA4f8R) upgraded the Share Sheet on Android 10 and above to show the title of the media/channel/playlist being shared. Some Android variants don't provide the "Copy to clipboard" option in the Share Sheet. So [@Abanoub8](https://github.com/Abanoub8) added the ability to long-press the Share button below the player, which copies the media URL directly into your clipboard.
```

Reviewer: thats a good one

[@TiA4f8R](https://github.com/TiA4f8R) also fixed video timestamps in the description so that they open in the popup player (just like timestamps in comments), instead of trying to open in another app entirely, or even in a restarted instance of NewPipe (which led to a loss of browsing session history/backstack).

[@B0pol](https://github.com/B0pol) updated the list of Invidious instance URLs that can be opened by NewPipe. He also added Piped URLs to the same list. For those who don't know, [Piped](https://piped.kavin.rocks/) is a great alternative YouTube front-end similar to Invidious, but based on the NewPipe Extractor instead! The biggest advantage to having this additional option is that if Invidious isn't working for you, Piped probably will, and vice-versa! Piped is developed by the awesome [@FireMasterK](https://github.com/FireMasterK), who you might recognise from previous blog posts and contributions to NewPipe.

> [@Stypox](https://github.com/Stypox) slightly improved the layout of streams in the Play Queue. #6443

Reviewer: ```suggestion
[@Stypox](https://github.com/Stypox) slightly improved the layout of streams in the Play Queue.
```

- Bump gson from 2.8.6 to 2.8.7 
- Bump rhino from 1.7.12 to 1.7.13 
- Bump spotbugs-annotations from 4.0.2 to 4.2.3

[@TobiGr](https://github.com/TobiGr) updated the hardcoded client ID for SoundCloud.

[@XiangRongLin](https://github.com/XiangRongLin) removed personal IPv4 addresses from the automatically generated mock tests and ensured that they won't slip in again.

Since NewPipe uses Java 8 now, [@Isira-Seneviratne](https://github.com/Isira-Seneviratne) went ahead and converted the abstract class DAOs (data access objects) to interfaces (Read: coder did some coding magic, now things are better somehow, don't ask us why).

[@Imericxu](https://github.com/Imericxu) improved the app code by resolving some Checkstyle warnings related to Tabs (Trending, Subscriptions, etc.).

[@Isira-Seneviratne](https://github.com/Isira-Seneviratne) updated the AndroidX Room library to v2.3.0. This also allowed him to migrate Converters (utility functions which provide easy conversion from one data type to another) to Kotlin.

> [@danielmbutler](https://github.com/danielmbutler) replaced the `System.exit` function calls used to close NewPipe with `getActivity.finishAffinity()`. Now instead of straight up killing the app process in one go, the OS will let the app close gracefully by letting it call all its functions to stop activities in the correct order, and free up memory gradually. This improvement was suggested by [@Redirion](https://github.com/Redirion).

Reviewer: ```suggestion
[@danielmbutler](https://github.com/danielmbutler) replaced the `System.exit` function calls used to close NewPipe with `getActivity.finishAffinity()`. Now instead of straight up killing the app process in one go, the OS will let the app close gracefully by allowing it to call all its functions to stop activities in the correct order, and free up memory gradually. This improvement was suggested by [@Redirion](https://github.com/Redirion).
```
